### PR TITLE
Feature/database refactor

### PR DESCRIPTION
**Pull request by mabiede**
- refactor multitenancy database interface
- handle disconnect and reconnection of tenant database
- integrate user, queue and tenant cache
- remove dependency on sihl database, sihl queue, sihl queue
- extract from pool common into separated modules: pool conformist, pool message, pool model
- add contact, admin and pool_user id

### CODE REVIEW

File: pool/queue/queue.ml [hunk 1]
```diff
@@ -1,163 +1,219 @@
...
+
+let registered_jobs : AnyJob.t list ref = ref []
+
+let register_jobs jobs =
+  registered_jobs := CCList.concat [ !registered_jobs; jobs ];
+  Lwt.return_unit
 ;;
 
-let registered_jobs : job' list ref = ref []
+(* let stop_schedule : (unit -> unit) option ref = ref None *)
```
Comment (timohuber) [inline pool/queue/queue.ml:16]: Plese remove

File: pool/queue/queue.ml [hunk 2]
```diff
@@ -170,19 +226,39 @@ let stop () =
...
+    ~implementation_name:"worker"
     ~dependencies:(fun () ->
-      [ Pool_canary.lifecycle; Database.lifecycle; Schedule.lifecycle ])
+      [ Pool_canary.lifecycle; Pool_database.lifecycle; Schedule.lifecycle ])
     ~start
     ~stop
 ;;
 
-let register ?(jobs = []) () =
+let register ?(kind = Service) ?(jobs = []) () =
```
Comment (timohuber) [inline pool/queue/queue.ml:252]: Is the kind `Worker` in use?

**Review by timohuber — CHANGES_REQUESTED** [review]

File: pool/app/pool_user/pool_user.ml [hunk 3]
```diff
@@ -1,32 +1,115 @@
...
+  include Entity_password
+
+  let validate_current label (user_id : Id.t) password =
+    Repo_password.find label user_id
+    ||> CCResult.to_opt
+    ||> CCOption.map_or ~default:false (CCFun.flip validate password)
+  ;;
+
+  let define label (user_id : Id.t) password password_confirmation =
+    let* (_ : t) = Repo_password.find label user_id in
```
[the remark this answers is not in the document]
Comment (mabiede) [reply pool/app/pool_user/pool_user.ml:16]: Check for the user's existence (which includes not "inactive"). As for "define", it's unnecessary to provide the old password.

File: pool/cqrs_command/contact_command.ml [hunk 4]
```diff
@@ -201,66 +200,61 @@ module UpdatePassword : sig
   include Common.CommandSig
```
[the remark this answers is not in the document]
Comment (mabiede) [reply pool/cqrs_command/contact_command.ml]: yes, done.

File: pool/app/assignment/repo/repo_entity.ml [hunk 5]
```diff
@@ -1,33 +1,42 @@
...
+  include Id
+
+  let t = make_caqti_type Caqti_type.string (of_string %> CCResult.return) value
+end
+
 module NoShow = struct
   include NoShow
 
-  let t = Caqti_type.bool
+  let t = make_caqti_type Caqti_type.bool (create %> CCResult.return) value
```
[the remark this answers is not in the document]
Comment (mabiede) [reply pool/app/assignment/repo/repo_entity.ml:15]: In an intermediate step, I added an entity interface file, which didn't expose the `type t`. I reset it.

File: pool/database/database.ml [hunk 6]
```diff
@@ -1,145 +1,133 @@
...
+
+  let start () =
+    log_start label;
+    let%lwt (_ : status) = setup () in
+    Lwt.return_unit
+  ;;
+
+  let stop () = Lwt.return_unit
+
+  let test_connection () =
```
[the remark this answers is not in the document]
Comment (mabiede) [reply pool/database/database.ml:54]: Yes, I was also thinking of adding a separate process to watch the connection to the root database. As it's not implemented yet, I removed it.

File: pool/database/entity.ml [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  module Core = struct
+    let print = Utils.ppx_printer
+    let field = Pool_message.Field.Status
+
+    type t =
+      | Active [@name "active"] [@printer print "active"]
+      | ConnectionIssue [@name "connection_issue"]
+      [@printer print "connection_issue"]
+      | Disabled [@name "disabled"] [@printer print "disabled"]
+      | Maintenance [@name "maintenance"] [@printer print "maintenance"]
```
[the remark this answers is not in the document]
Comment (mabiede) [reply pool/database/entity.ml:48]: The database and the tenant share the status, as they cannot exist without each other.

File: pool/database/pools.ml [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+  type status =
+    ( (Caqti_lwt.connection, Caqti_error.t) Caqti_lwt_unix.Pool.t
+      , Caqti_error.load )
+      result
+
+  module Config = Config
+  module Hashtbl = CCHashtbl.Make (Label)
+
+  let pools : status Hashtbl.t =
+    Hashtbl.create (max 1 Config.expected_databases)
```
[the remark this answers is not in the document]
Comment (mabiede) [reply pool/database/pools.ml:48]: This is to make sure the Hashtable gets initialized with a natural size. Basically, it's possible to set the "expected_databases" configuration to a negative number.

File: pool/database/service.ml [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+  "SET FOREIGN_KEY_CHECKS = ?" |> Caqti_type.(bool ->. unit)
+;;
+
+let with_disabled_fk_check database_label f =
+  let open Lwt_result.Syntax in
+  query database_label (fun connection ->
+    let module Connection = (val connection : Caqti_lwt.CONNECTION) in
+    let* () = Connection.exec set_fk_check_request false in
+    (f connection)
+      (* Use PPX for backtrace *)
```
[the remark this answers is not in the document]
Comment (mabiede) [reply pool/database/service.ml:60]: No, this comment belongs to the line below (intended).

File: pool/queue/queue.ml (hunk 2 above)
Comment (mabiede) [reply pool/queue/queue.ml:252]: Yes, for the "worker" ;-)

File: pool/database/pools.ml [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+          let result = result |> CCResult.get_exn in
+          match%lwt Connection.commit () with
+          | Ok () -> Lwt.return_ok result
+          | Error error -> Lwt.return_error error)
+        (rollback connection))
+    |> map_fetched label
+    ||> raise_caqti_error
+  ;;
+
+  let transactions label queries =
```
[the remark this answers is not in the document]
Comment (timohuber) [reply pool/database/pools.ml:222]: as discussed, lets update naming

**Review by timohuber — APPROVED** [review]

**Pull request merged**